File: 136_single_number/singleNumber.go
package singlenumber

func singleNumber(nums []int) int {
	unique := getUnique(nums)
	currSum := sum(nums)
	uniqueSum := sum(unique)

	return uniqueSum*2 - currSum
}

func sum(nums []int) int {
	s := 0

	for _, num := range nums {
		s += num
	}

	return s
}

func getUnique(nums []int) []int {
	res := make([]int, 0)
	set := make(map[int]bool)

	for _, num := range nums {
		if _, ok := set[num]; !ok {
			res = append(res, num)
		}

		set[num] = true
	}

	return res
}

File: 136_single_number/single_number.rs
use std::collections::HashMap;

impl Solution {
    pub fn single_number(nums: Vec<i32>) -> i32 {
        let unique = get_unique(&nums);
        let curr_sum = sum(&nums);
        let unique_sum = sum(&unique);
        
        unique_sum * 2 - curr_sum
    }
}

fn sum(nums: &Vec<i32>) -> i32 {
    let mut sum = 0;

    for &v in nums {
        sum += v;
    }

    sum
}

fn get_unique(nums: &Vec<i32>) -> Vec<i32> {
        let mut h_map = HashMap::new();
        let mut res = Vec::new();

        for &v in nums {
            match h_map.get(&v) {
                Some(_) => (),
                _ => {
                    h_map.insert(v, true);
                    res.push(v);
                }
            }
        }

        res
}